lenskit-svd/src/main/java/org/lenskit/mf/BPR/BPRMFModelProvider.java

```diff
@@ -0,0 +1,175 @@
+        RollingWindowMeanAccumulator optAccum = new RollingWindowMeanAccumulator(1000, -30);
+
+        int userCount = dao.getEntityIds(CommonTypes.USER).size();
+        RealMatrix userFeatures = MatrixUtils.createRealMatrix(userCount, featureCount);
+        initFeatures(userFeatures);
+
+        int itemCount = dao.getEntityIds(CommonTypes.ITEM).size();
+        RealMatrix itemFeatures = MatrixUtils.createRealMatrix(itemCount, featureCount);
+        initFeatures(userFeatures);
+
+        logger.debug("Learning rate is {}", learningRate);
+        logger.debug("Regularization term is {}", regularization);
+
+        logger.info("Building MPR-MF with {} features for {} users and {} items",
+                featureCount, userCount, itemCount);
+
+        TrainingLoopController controller = stoppingCondition.newLoop();
+
+        //REVIEW: because of the nature of training samples (and the point that the BPR paper makes that training
+        // by-item or by-user are not optimal) one "iteration" here will be one training update. This leads to _really_
+        // big iteration counts, which can actually overflow ints!. one suggestion would be to allow the iteration count
+        // controller to accept longs, but I don't know if this will introduce backwards compatibility issues (I imagine
+        // this depends on the robustness of our type conversion in the configuration.
+        while(controller.keepTraining(-optAccum.getMean())) {
+            TrainingItemPair pair = pairGenerator.nextPair();
+            // Note: bad code style variable names are generally to match BPR paper and enable easier implementation
+            long iid = pair.g;
+            int i = itemIndex.internId(iid);
+            long jid = pair.l;
+            int j = itemIndex.internId(jid);
+            long uid = pair.u;
+            int u = userIndex.internId(uid);
+
+            RealVector w_u = userFeatures.getRowVector(u);
+            RealVector h_i = itemFeatures.getRowVector(i);
+            RealVector h_j = itemFeatures.getRowVector(j);
+
+            double xui = w_u.dotProduct(h_i);
+            double xuj = w_u.dotProduct(h_j);
+            double xuij = xui - xuj;
+
+            double bprTerm = 1/(1+exp(xuij));
+
+            // w_u update
+            RealVector h_i_j = h_i.subtract(h_j);
+            RealVector w_u_update = w_u.mapMultiply(-regularization);
+            w_u_update.combineToSelf(1, bprTerm, h_i_j);
+
+            // h_i update
+            RealVector h_i_update = h_i.mapMultiply(-regularization);
+            h_i_update.combineToSelf(1, bprTerm, w_u);
+
+            // h_j update
+            RealVector h_j_update = h_j.mapMultiply(-regularization);
+            h_j_update.combineToSelf(1, -bprTerm, w_u);
+
+            // perform updates
+            w_u.combineToSelf(1, learningRate, w_u_update);
+            h_i.combineToSelf(1, learningRate, h_i_update);
+            h_j.combineToSelf(1, learningRate, h_j_update);
```

Comment:
This won't actually update without calls to the `setRowVector` method on the matrix..